homeassistant/components/history/__init__.py

```diff
@@ -300,13 +362,409 @@ def _sorted_significant_states_json(
+        if events[0].time_fired <= subscriptions_setup_complete_time:
+            continue
+        # We sleep for the EVENT_COALESCE_TIME so
+        # we can group events together to minimize
+        # the number of websocket messages when the
+        # system is overloaded with an event storm
+        await asyncio.sleep(EVENT_COALESCE_TIME)
+        while not stream_queue.empty():
+            events.append(stream_queue.get_nowait())
+
+        if history_states := _events_to_compressed_states(events, no_attributes):
+            connection.send_message(
+                JSON_DUMP(
+                    messages.event_message(
+                        msg_id,
+                        {"states": history_states},
+                    )
+                )
+            )
+
+
+@callback
+def _async_subscribe_events(
+    hass: HomeAssistant,
+    subscriptions: list[CALLBACK_TYPE],
+    target: Callable[[Event], None],
+    entities_filter: EntityFilter | None,
+    entity_ids: list[str] | None,
+    significant_changes_only: bool,
+    minimal_response: bool,
+) -> None:
+    """Subscribe to events for the entities and devices or all.
+
+    These are the events we need to listen for to do
+    the live history stream.
+    """
+    assert is_callback(target), "target must be a callback"
+
+    @callback
+    def _forward_state_events_filtered(event: Event) -> None:
+        """Filter state events and forward them."""
+        if (new_state := event.data.get("new_state")) is None or (
+            old_state := event.data.get("old_state")
+        ) is None:
+            return
+        assert isinstance(new_state, State)
+        assert isinstance(old_state, State)
+        if (entities_filter and not entities_filter(new_state.entity_id)) or (
+            (significant_changes_only or minimal_response)
+            and new_state.state == old_state.state
+            and new_state.domain not in history.SIGNIFICANT_DOMAINS
+        ):
+            return
+        target(event)
+
+    if entity_ids:
+        subscriptions.append(
+            async_track_state_change_event(
+                hass, entity_ids, _forward_state_events_filtered
+            )
+        )
+        return
+
+    # We want the firehose
+    subscriptions.append(
+        hass.bus.async_listen(
+            EVENT_STATE_CHANGED,
+            _forward_state_events_filtered,
+            run_immediately=True,
+        )
+    )
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "history/stream",
+        vol.Required("start_time"): str,
+        vol.Optional("end_time"): str,
+        vol.Optional("entity_ids"): [str],
+        vol.Optional("include_start_time_state", default=True): bool,
+        vol.Optional("significant_changes_only", default=True): bool,
+        vol.Optional("minimal_response", default=False): bool,
+        vol.Optional("no_attributes", default=False): bool,
+    }
+)
+@websocket_api.async_response
+async def ws_stream(
```

Comment:
Given the amount of code that is involved for the streaming, I think it would make sense to move this to a `websocket_api.py` command.

Reply:
100% agree. Was planning on doing that in a followup.

Reply:
Split in https://github.com/home-assistant/core/pull/86364